boot/flags.go

```diff
@@ -280,3 +282,60 @@ func setNextBootFlags(dev Device, rootDir string, flags []string) error {
 
 	return m.Write()
 }
+
+// RunModeRootfs returns where the run mode root filesystem is mounted for the
+// given mode.
+// For run mode, it's just "/run/mnt/data".
+// For install mode it's "/run/mnt/ubuntu-data".
+// For recover mode it's either "/host/ubuntu-data" or "" if that is not
+// mounted. Note that, for recover  mode, this function only returns a non-empty
+// return value if the partition is mounted and trusted, there are certain
+// corner-cases where snap-bootstrap in the initramfs may have mounted
+// ubuntu-data in an untrusted manner, but for the purposes of this function
+// that is ignored.
+// This is primarily meant to be consumed by "snap{,ctl} system-mode".
+func RunModeRootfs(mode string) (string, error) {
+	runDataRootfs := ""
+	switch mode {
+	case ModeRun:
+		runDataRootfs = InitramfsDataDir
+	case ModeRecover:
+		// TODO: should this be it's own dedicated helper to read degraded.json?
+
+		// for recover mode, the source of truth to determine if we have the
+		// host mount is snap-bootstrap's /run/snapd/snap-bootstrap/degraded.json, so
+		// we have to go parse that
+		degradedJSONFile := filepath.Join(dirs.SnapBootstrapRunDir, "degraded.json")
+		b, err := ioutil.ReadFile(degradedJSONFile)
+		if err != nil {
+			return "", err
+		}
+
+		degradedJSON := struct {
+			UbuntuData struct {
+				MountState    string `json:"mount-state"`
+				MountLocation string `json:"mount-location"`
+			} `json:"ubuntu-data"`
+		}{}
+
+		err = json.Unmarshal(b, &degradedJSON)
+		if err != nil {
+			return "", err
+		}
+
+		// don't permit mounted-untrusted state, only mounted state is allowed
+		if degradedJSON.UbuntuData.MountState == "mounted" {
+			runDataRootfs = degradedJSON.UbuntuData.MountLocation
+		}
+		// otherwise leave it empty
+
+	case ModeInstall:
+		// this is /run/mnt/ubuntu-data/writable, but the caller probably wants
+		// /run/mnt/ubuntu-data
+		runDataRootfs = filepath.Dir(InstallHostWritableDir)
```

Comment:
this is only true if we are past data partition creation, basically we should maybe check if something is there and otherwise return ""

Reply:
I guess I was implicitly assuming that this would only ever be used from the install-device hook during install mode, which is only invoked if partition creation was successful, but that's a good point we could maybe some day use this elsewhere

Reply:
I made it so we check that the dir exists before returning it here